### PR TITLE
Adding sugarscape simulation

### PR DESCRIPTION
**Pull request by chelsea0x3b**

### CODE REVIEW

File: examples/sugarscape.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+        # self.sugar = self.max_sugar
+
+        # constant growback model
+        self.sugar = min(self.sugar + 1, self.max_sugar)
+
+    def __repr__(self):
+        return 'Patch({}, {})'.format(self.sugar, self.max_sugar)
+
+
+class Turtle(dworp.SelfNamingAgent):
```
Comment (cash) [inline examples/sugarscape.py]: Turtle is the name NetLogo gives to all agents. It's not specific to this paper. NetLogo is based on [Logo](https://en.wikipedia.org/wiki/Logo_(programming_language)) (the first computer language that I learned when I was 7). It had turtle graphics where you had this triangle on the screen and could issue commands like move or turn. As it moved it drew lines on the screen. It looked something like this: https://www.youtube.com/watch?v=mttyoaspa7U

Anyway, we should come up with a better name than turtle for the agents. Maybe SugarAgent

File: examples/sugarscape.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+#
+# To run, you must install pygame:
+# pip install pygame
+#
+# Original paper:
+# Epstein, J.M. & Axtell, R. Artificial Life and Robotics (1997) 1: 33. https://doi.org/10.1007/BF02471109
+#
+# This reproduces the model according to NetLogo.
+#
+# The pygame visualization seems to have a busy wait loop which will peg a processor.
```
Comment (cash) [inline examples/sugarscape.py:11]: I looked into that. If I remember correctly pygame did have a busy loop.

**Review by cash — COMMENTED** [review]
Are you seeing similar behavior as the NetLogo sim? For example, some of the agents get caught on the edge of the sugar and seem to just sit there.

File: examples/sugarscape.py (hunk 2 above)
Comment (chelsea0x3b) [reply examples/sugarscape.py:11]: Ahh okay, that comment was copied from one of the other examples, should I change it?

**Comment by chelsea0x3b** [comment]
Yes seeing similar behavior now.

Here's the vis in the first variant (immediate growback):
dworp-sugarscape-immediate

Here's the vis in the second variant (constant growback):
dworp-sugarscape-constant

After looking at the netlogo vis, I noticed that in the second variant, their agents were wandering randomly. That's where the rng commit I just added comes in.

File: examples/sugarscape.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+        for agent in agents:
+            empty_locations = env.empty_locations()
+
+            # note: can't just do self.rng.choice(empty_locations), because numpy thinks a list of tuples is
+            # multi-dimensional, and rng.choice only works with 1-d arrays.
+            env.move(agent, empty_locations[self.rng.choice(range(len(empty_locations)))])
+
+        super().__init__(agents, env, dworp.InfiniteTime(), dworp.BasicScheduler(), observer, SugarscapeTerminator())
+
+    def create_turtle(self):
```
Comment (cash) [inline examples/sugarscape.py]: update the reference to turtle

**Pull request merged**